Feature/singularity addition

EdwardSnyder-NOAA opened the pull request:
## DESCRIPTION OF CHANGES: 
Added singularity build and workflow environment scripts.

## TESTS CONDUCTED: 
Test was conducted on Orion and Parallel Works cloud environments. 

## DEPENDENCIES:
None but it is related to the regional workflow[ PR #686](https://github.com/ufs-community/regional_workflow/pull/686/). 

## DOCUMENTATION:
Future PR will capture these changes regarding the use of the singularity container. Not changing model configuration. 

## ISSUE (optional): 

## CONTRIBUTORS (optional): 
@jkbk2004 @mark-a-potts

EdwardSnyder-NOAA committed:
Create build_singularity_gnu.env

EdwardSnyder-NOAA committed:
Create wflow_singularity.env

jkbk2004 committed:
Updated regional_workflow pointer: ES fork branch

EdwardSnyder-NOAA committed:
Merge pull request #1 from jkbk2004/patch-1

Updated regional_workflow pointer: ES fork branch

christopherwharrop-noaa commented on env/build_singularity_gnu.env:
```diff
@@ -0,0 +1,16 @@
+
+module load hpc hpc-gnu hpc-openmpi
+
+module load netcdf hdf5 bacio sfcio sigio nemsio w3emc esmf fms crtm g2 png zlib g2tmpl ip sp w3nco cmake/3.21.1 gfsio wgrib2 upp
```
Is there a reason this is not doing a `module load srw_common` like the rest of the env files?

jkbk2004 replied:
@christopherwharrop-noaa We are trying to build the containers: still frequent update. I think we can adopt srw_common  soon on singularity but it needs time. @EdwardSnyder-NOAA can you add a comment line about the container url in the singularity env file? just for reference but details will be available at docs.

EdwardSnyder-NOAA replied:
Added comment in the singularity env file that mentions how the singularity image is built and how to find that docker image.

christopherwharrop-noaa reviewed (commented):
I'm not quite understanding how this is a singularity build?  I was expecting to see a Singularity definition file.  Can you explain a bit more about how this will work?

EdwardSnyder-NOAA committed:
Update build_singularity_gnu.env

christinaholtNOAA reviewed (commented):
This is redundant, but I left the same comment on @mark-a-potts PR #221. I think it is even more relevant as we are talking about adding both Singularity as a platform and Parallel Works.

I have a broader question for the group about how to maintain ongoing support for cloud platforms. Given that there is essentially no governance over how a cloud platform is created (naming conventions of lustre file systems, partitions, locations of software, etc.), I'd like to hear thoughts on how we can both support cloud users, and make sure that we aren't tying them down to a restrictive set of requirements for their platform.

Note: I am one of those cloud users, and am quite interested in this support!

louiswicker commented:
C - 

I guess my comment on the cloud support would favor a container approach - as I think the support can then be consolidated.  I was under the impression (though I could be wrong) that containers were going to be supported on Parallel works?

Lou

> On Mar 4, 2022, at 9:56 AM, Christina Holt ***@***.***> wrote:
> 
> 
> @christinaholtNOAA commented on this pull request.
> 
> This is redundant, but I left the same comment on @mark-a-potts <https://github.com/mark-a-potts> PR #221 <https://github.com/ufs-community/ufs-srweather-app/pull/221>. I think it is even more relevant as we are talking about adding both Singularity as a platform and Parallel Works.
> 
> I have a broader question for the group about how to maintain ongoing support for cloud platforms. Given that there is essentially no governance over how a cloud platform is created (naming conventions of lustre file systems, partitions, locations of software, etc.), I'd like to hear thoughts on how we can both support cloud users, and make sure that we aren't tying them down to a restrictive set of requirements for their platform.
> 
> Note: I am one of those cloud users, and am quite interested in this support!
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/ufs-community/ufs-srweather-app/pull/220#pullrequestreview-900401154>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AADUKPS3M5DRJERWOMHQU5TU6IXETANCNFSM5PYNO3MQ>.
> Triage notifications on the go with GitHub Mobile for iOS <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>. 
> You are receiving this because you are subscribed to this thread.
> 

----------------------------------------------------------------------------
| Dr. Louis J. Wicker
| (he/him/his)
| NSSL/FRDD  Rm 3336
| National Weather Center
| 120 David L. Boren Boulevard, Norman, OK 73072
|
| E-mail:   ***@***.***
| HTTP:    www.nssl.noaa.gov/~lwicker
| Phone:    (405) 325-6340
| Fax:        (405) 325-2316
|
|
| Just as the constant increase of entropy is the
| basic law of the universe, 
| so it is the basic law of life to be ever more highly 
| structured and to struggle against entropy.
| 
|  Vaclav Havel
----------------------------------------------------------------------------
|
| "The contents  of this message are mine personally and 
| do not reflect any position of  the Government or NOAA."
|
----------------------------------------------------------------------------

mark-a-potts commented:
> This is redundant, but I left the same comment on @mark-a-potts PR #221. I think it is even more relevant as we are talking about adding both Singularity as a platform and Parallel Works.
> 
> I have a broader question for the group about how to maintain ongoing support for cloud platforms. Given that there is essentially no governance over how a cloud platform is created (naming conventions of lustre file systems, partitions, locations of software, etc.), I'd like to hear thoughts on how we can both support cloud users, and make sure that we aren't tying them down to a restrictive set of requirements for their platform.
> 
> Note: I am one of those cloud users, and am quite interested in this support!

So, the NOAACLOUD platform that is referenced in PR#221 is specific to the cloud instances available to NOAA users through the ParallelWorks dashboard.  All three providers (AWS, Azure, and Google) differ slightly, but have the same directory structure and their software stacks are very similar. 

Ideally, offering a container-based version will allow users to just move that to any platform (cloud or not) and start using the SRW. Unless I am mistaken, this PR is a step in that direction, but is still restricted to running on a single node. It is possible to run MPI executables like the model or the post across multiple nodes, but doing so as part of the current workflow is not really feasible. This is due to the basic structure of "load_modules_and_run.sh". In order to run across nodes, you want to run (call mpiexec/mpirun/srun) from outside the container and set the environment variable inside the container. It might be possible to convert all the required environment variables into something that gets passed into singularity (SINGULARITYENV_LD_LIBRARY_PATH would get picked up inside singularity as just LD_LIBRARY_PATH, for example), but it would definitely take some work, and there is probably a better, more elegant approach.

dshawul commented:
We have done something similar with global_workflow,  I agree it is a mess though doable.
Having both rocoto and slurm sitting outside of the container, and binaries in the container, you have to go back and forth in/outside of the container. Workflow scripts need to modified to do `singularity exec CONTAINER /path/to/exec`, environement variables need to be passed around to/from container etc... OTOH though the current approach is limited to one node that makes it not usable for parts of the workflow that need multiple nodes.

danielabdi-noaa commented:
We have done something similar with global_workflow, I agree it is a mess though doable.
Having both rocoto and slurm sitting outside of the container, and binaries in the container, you have to go back and forth in/outside of the container. Workflow scripts need to modified to do `singularity exec CONTAINER /path/to/exec`, environement variables need to be passed around to/from container etc... OTOH though the current approach is limited to one node that makes it not usable for parts of the workflow that need multiple nodes.

EdwardSnyder-NOAA committed:
Update Externals.cfg

Updated the Regional Workflow repo as the accompanied PR (https://github.com/ufs-community/regional_workflow/pull/686) has been merged into the develop branch.

EdwardSnyder-NOAA reviewed (commented):
updated regional workflow repo url

jkbk2004 commented:
@EdwardSnyder-NOAA from the comments, certainly there is a technical barrier between container and local resource management. But I agree with @mark-a-potts merging this PR will allow us to explore more container-based options. I hope to see merging the PR soon as we tested ok on script-based and single node approaches though. I am ready to approve the PR. Also, detail documentation is following up about the approaches, right?

jkbk2004 reviewed (approved):
This PR is pretty much isolated case and tested ok with container. So I am approving. Thanks fpr the work, @EdwardSnyder-NOAA

JeffBeck-NOAA commented on Externals.cfg:
```diff
@@ -2,8 +2,8 @@
-hash = e5d5c38
+branch = develop
+#hash = e5d5c38
 local_path = regional_workflow
```
I assume a hash update commit will be pushed here once the corresponding regional_workflow [PR](https://github.com/ufs-community/regional_workflow/pull/686) is merged?

EdwardSnyder-NOAA replied:
Updated the regional workflow hash.

jkbk2004 commented:
@JeffBeck-NOAA @EdwardSnyder-NOAA sorry I was a bit confused about merging process here. We actually merged in the regional_workflow PR Friday. So the regional workflow develop branch contains the singularity option in it.

JeffBeck-NOAA commented:
@jkbk2004, not a problem.  In that case, this PR needs to have the regional_workflow hash updated in Externals.cfg to what is currently being used for the HEAD.

EdwardSnyder-NOAA committed:
Update Externals.cfg

updated regional workflow hash

christinaholtNOAA commented on env/build_singularity_gnu.env:
```diff
@@ -0,0 +1,17 @@
+
+module use /opt/hpc-modules/modulefiles/stack
+
+module load hpc hpc-gnu hpc-openmpi
```
It is often easier to manage lists of modules when they are one-per-line. Would you mind making that change?

EdwardSnyder-NOAA committed:
Update build_singularity_gnu.env

christinaholtNOAA reviewed (approved):
Thanks!
